### DISCUSSION TITLE
model quality parameters on a remote machine?

### DISCUSSION THREAD

**Discussion by eringiglio** [talmolab/sleap#1424] [Help!] [answered]
Hi! I have a really big dataset (~20tb) and spend a lot of time working on a remote cluster system, and my current workflow involves running models that I'm trying to train to recognize a single mouse in an open field across ~33 90min videos apiece. It's therefore really important to me to be able to compare model parameters to make sure that I'm making headway on this enormous dataset. 

However, I'm having a lot of trouble running command-line assessments of model quality on my computing cluster. It seems to be a pathing error: even though I'm trying to run lines like this with a .pkg.slp file generated from a training job, which should theoretically contain all its images locally, the program keeps trying to pull frames from the source videos--which are, of course, totally unavailable since the paths were set on a Windows machine and I'm now trying to work on a Linux headless environment. It's very frustrating. I have successfully run many of these model evaluations locally on the same Windows machine where the .pkg.slp files were generated, but I'm working at the limit of my poor laptop's capacity and it means I'm moving very slowly when I try to compare models on the same labels to one another. (We're talking 1000-4000 labeled frames per project here; it's not trivial!) 

Is there anything I can do about this, like generating a .pkg.slp file from a .slp file tuned to the file paths for the linux environment from the command line? How can I compare model performance on the same sets of frames given the remote environment? I've tried setting paths manually to the correct video files by editing the videos' source_video_available.backend.filename, but it hasn't been successful. I really want to be able to compare these outputs so I can improve the outcomes effectively and get digging into my analysis pipeline! 

Let me know if example files or code would be helpful.

**Comment by eringiglio** [comment]
As I dig into this, notes in case others find them helpful...

A sample of error outputs.

> 2023-07-31 12:32:20.347312: W tensorflow/core/framework/op_kernel.cc:1680] Unknown: IndexError: Frame index 107285 not in original index.Traceback (most recent call last):
> 
>   File "/home/ngrissom/gigli029/.conda/envs/sleap/lib/python3.7/site-packages/sleap/io/video.py", line 239, in _try_frame_from_source_video
>     return self.source_video.get_frame(idx)
> 
>   File "/home/ngrissom/gigli029/.conda/envs/sleap/lib/python3.7/site-packages/sleap/io/video.py", line 204, in source_video
>     if self.source_video_available:
> 
>   File "/home/ngrissom/gigli029/.conda/envs/sleap/lib/python3.7/site-packages/sleap/io/video.py", line 1000, in __len__
>     return self.frames
> 
>   File "/home/ngrissom/gigli029/.conda/envs/sleap/lib/python3.7/site-packages/sleap/io/video.py", line 964, in __getattr__
>     return getattr(self.backend, item)
> 
>   File "/home/ngrissom/gigli029/.conda/envs/sleap/lib/python3.7/site-packages/sleap/io/video.py", line 439, in frames
>     return int(self.__frames_float)
> 
>   File "/home/ngrissom/gigli029/.conda/envs/sleap/lib/python3.7/site-packages/sleap/io/video.py", line 400, in __frames_float
>     return self.__reader.get(cv2.CAP_PROP_FRAME_COUNT)
> 
>   File "/home/ngrissom/gigli029/.conda/envs/sleap/lib/python3.7/site-packages/sleap/io/video.py", line 381, in __reader
>     f"Could not find filename video filename named {self.filename}"
> 
> FileNotFoundError: Could not find filename video filename named D:/hash2/hash2_drug2.16.MP4
> 
> 
> During handling of the above exception, another exception occurred:
> 
> 
> Traceback (most recent call last):
> 
>   File "/home/ngrissom/gigli029/.conda/envs/sleap/lib/python3.7/site-packages/tensorflow/python/ops/script_ops.py", line 247, in __call__    return func(device, token, args)
> 
>   File "/home/ngrissom/gigli029/.conda/envs/sleap/lib/python3.7/site-packages/tensorflow/python/ops/script_ops.py", line 135, in __call__    ret = self._func(*args)
> 
>   File "/home/ngrissom/gigli029/.conda/envs/sleap/lib/python3.7/site-packages/tensorflow/python/autograph/impl/api.py", line 645, in wrapper
>     return func(*args, **kwargs)
> 
>   File "/home/ngrissom/gigli029/.conda/envs/sleap/lib/python3.7/site-packages/sleap/nn/data/providers.py", line 198, in py_fetch_lf      
>     raw_image = lf.image
> 
>   File "/home/ngrissom/gigli029/.conda/envs/sleap/lib/python3.7/site-packages/sleap/instance.py", line 1757, in image
>     return self.video.get_frame(self.frame_idx)
> 
>   File "/home/ngrissom/gigli029/.conda/envs/sleap/lib/python3.7/site-packages/sleap/io/video.py", line 1022, in get_frame
>     return self.backend.get_frame(idx)
> 
>   File "/home/ngrissom/gigli029/.conda/envs/sleap/lib/python3.7/site-packages/sleap/io/video.py", line 317, in get_frame
>     return self._try_frame_from_source_video(idx)
> 
>   File "/home/ngrissom/gigli029/.conda/envs/sleap/lib/python3.7/site-packages/sleap/io/video.py", line 241, in _try_frame_from_source_video
>     raise IndexError(f"Frame index {idx} not in original index.")
> 
> IndexError: Frame index 107285 not in original index.

So the problem is clearly a case of "What happens if the image for a particular frame that _should_ be enclosed in a .pkg.slp file isn't there?" 

Currently trying out seeing if I can set the value for each video's source_video_available to False, seeing what that does.

**Comment by eringiglio** [comment]
Nope, not useful. 

Also not useful: creating a pkg.slp file with just labeled frames from the original main project file and trying with that. 

Next up: going to try changing file paths in the main project file and export from there. Fingers crossed.

**Comment by eringiglio** [comment, marked as answer]
SUCCESS. 

Final pipeline involves copying the main .slp project file, changing all the file paths to point to files hosted on the remote cluster, and exporting the user-labeled frames to a .pkg.slp file also hosted locally on the cluster. That works well. 

example code for the sake of any other poor fuckers working with data on this scale: 

```
    slp = sleap.load_file(os.path.join(mouseDir,mouse+'.slp'))
    slpkgPath = os.path.join(mouseDir,mouse+'.pkg.slp')
    for video in slp.videos:
        oldName = video.filename
        newName = oldName.replace('D:/','/scratch.global/gigli029/breakfast/males/final_videos/')
        video.backend.filename = newName
        video.filename = newName
    labels = sleap.Labels(slp.user_labeled_frames)  # all I care about here are user-labeled frames 
    labels.save(slpkgPath, with_images=True, embed_all_labeled=True)
```

**Reply by talmo** [reply]
Hi @eringiglio,

Thanks for following up with a solution!! Apologies that we weren't able to get to it in time!

This looks like a reasonable approach. I would just make sure that you're still able to import the results back (e.g., predictions on suggestions) after training with this package file. If the paths are off, it might create duplicate videos when you try to merge it back in.

Of course, if you just want the trained model and you'll re-do the inference locally, it shouldn't matter.

Cheers,

Talmo
